File: typescript/packages/var-core/tests/failure.test.ts
import { expect, test } from 'vitest'
import { CellMismatchError, ReturnShapeError } from '../src/cell-diff.ts'
import { DocStringMismatchError } from '../src/doc-string-diff.ts'
import { toFailure } from '../src/failure.ts'
import { spanFromOffsets } from '../src/span.ts'

test('toFailure extracts cells from a CellMismatchError', () => {
  const source = 'a | 5 |'
  const err = new CellMismatchError([
    { column: 'n', span: spanFromOffsets(source, 4, 5), expected: '5', actual: '4', ok: false },
  ])
  const f = toFailure(err, 'spec.md', 3)
  expect(f.cells).toEqual([{ from: 4, to: 5, actual: '4' }])
  expect(f.doc).toBeUndefined()
  expect(typeof f.message).toBe('string')
  expect(typeof f.stack).toBe('string')
})

test('toFailure extracts doc from a DocStringMismatchError', () => {
  const source = 'Hello!\n'
  const err = new DocStringMismatchError({
    span: spanFromOffsets(source, 0, 7),
    expected: 'Hello!\n',
    actual: 'Goodbye!\n',
  })
  const f = toFailure(err, 'spec.md', 3)
  expect(f.doc).toEqual({ from: 0, to: 7, actual: 'Goodbye!\n' })
  expect(f.cells).toBeUndefined()
})

test('toFailure leaves cells/doc undefined for a plain error or ReturnShapeError', () => {
  expect(toFailure(new Error('nope'), 'spec.md', 3).cells).toBeUndefined()
  expect(toFailure(new Error('nope'), 'spec.md', 3).doc).toBeUndefined()
  expect(toFailure(new ReturnShapeError('bad'), 'spec.md', 3).cells).toBeUndefined()
})

test('toFailure reads the failing line from an injected stack frame, else falls back', () => {
  const err = new Error('boom')
  err.stack = 'Error: boom\n    at handler (steps.ts:1:1)\n    at step (docs/a.md:12:3)'
  expect(toFailure(err, 'docs/a.md', 99).line).toBe(12)

  const noFrame = new Error('boom')
  noFrame.stack = 'Error: boom\n    at handler (steps.ts:1:1)'
  expect(toFailure(noFrame, 'docs/a.md', 99).line).toBe(99)
})

test('toFailure regex-escapes the spec path (a dot is literal)', () => {
  const err = new Error('boom')
  // 'X' stands in for the dot: if the spec path's `.` were treated as a regex
  // wildcard it would match this frame; escaped, it must not.
  err.stack = 'Error: boom\n    at step (aXmd:7:1)'
  // specPath "a.md" must NOT match "aXmd"
  expect(toFailure(err, 'a.md', 42).line).toBe(42)
})

File: java/core/src/test/java/dev/varar/core/FailureTest.java
package dev.varar.core;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertNull;

import java.util.List;
import org.junit.jupiter.api.Test;

/**
 * Translated from {@code var-core/tests/failure.test.ts}, adapted for Java's structured {@link
 * StackTraceElement} stack (see {@code Failure.java}'s class javadoc for why): the TS test
 * manually assigns a text {@code err.stack}; here a synthetic {@link StackTraceElement} plays
 * the same role as the frame {@code execute.ts}'s {@code augmentStack} would inject.
 */
class FailureTest {

    @Test
    void toFailureExtractsCellsFromACellMismatchException() {
        String source = "a | 5 |";
        CellDiff.CellMismatchException err = new CellDiff.CellMismatchException(
                List.of(new CellDiff("n", Span.spanFromOffsets(source, 4, 5), "5", "4", false)));
        Result.ExampleFailure f = Failure.toFailure(err, "spec.md", 3);
        assertEquals(List.of(new Result.CellFailure(4, 5, "4")), f.cells());
        assertEquals(String.class, f.message().getClass());
        assertEquals(String.class, f.stack().getClass());
    }

    @Test
    void toFailureExtractsADocStringMismatchAsACell() {
        String source = "Hello!\n";
        CellDiff diff = DocStringDiff.compareDocString("Goodbye!\n", "Hello!\n", Span.spanFromOffsets(source, 0, 7));
        Result.ExampleFailure f =
                Failure.toFailure(new CellDiff.CellMismatchException(java.util.List.of(diff)), "spec.md", 3);
        assertEquals(java.util.List.of(new Result.CellFailure(0, 7, "\"Goodbye!\\n\"")), f.cells());
    }

    @Test
    void toFailureLeavesCellsNullForAPlainExceptionOrReturnShapeException() {
        assertNull(Failure.toFailure(new RuntimeException("nope"), "spec.md", 3).cells());
        assertNull(Failure.toFailure(new CellDiff.ReturnShapeException("bad"), "spec.md", 3)
                .cells());
    }

    @Test
    void toFailureReadsTheFailingLineFromAnInjectedStackFrameElseFallsBack() {
        RuntimeException err = new RuntimeException("boom");
        err.setStackTrace(new StackTraceElement[] {
            new StackTraceElement("Handler", "handle", "steps.java", 1),
            new StackTraceElement("Step", "run", "docs/a.md", 12)
        });
        assertEquals(12, Failure.toFailure(err, "docs/a.md", 99).line());

        RuntimeException noFrame = new RuntimeException("boom");
        noFrame.setStackTrace(new StackTraceElement[] {new StackTraceElement("Handler", "handle", "steps.java", 1)});
        assertEquals(99, Failure.toFailure(noFrame, "docs/a.md", 99).line());
    }

    @Test
    void toFailureRegexEscapesTheSpecPathADotIsLiteral() {
        RuntimeException err = new RuntimeException("boom");
        // 'X' stands in for the dot: if the spec path's `.` were treated as a regex wildcard it
        // would match this frame; escaped, it must not.
        err.setStackTrace(new StackTraceElement[] {new StackTraceElement("Step", "run", "aXmd", 7)});
        // specPath "a.md" must NOT match "aXmd".
        assertEquals(42, Failure.toFailure(err, "a.md", 42).line());
    }
}
